Use filename property instead of generating a new filename + remove duplicate code for PDF invoice formatting

JoryHogeveen opened the pull request:
Prevents inconsistency with overrides.
After extra commits:
- Refer to the PDF filename instead of using duplicate code for PDF filename.
- Reuse OrderInvoice code to get PDF filename. This makes sure the invoice formatting logic is only done in one single place: the OrderInvoice class.

<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
------------------------------------------------------------------------------>

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | We used a override on the invoice number. Unfortunately the filename and invoice number is hardcoded in some locations. This PR fixes one of these.
| Type?             | improvement
| Category?         | CO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #26646.
| How to test?      | Override `getInvoiceNumberFormatted()` in `order/OrderInvoice.php` and that override should work in all locations where invoice numbers are used, even PDF filenames.
| Possible impacts? | None


<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/26587)
<!-- Reviewable:end -->

JoryHogeveen committed:
Use filename property instead of generating a new filename

matks reviewed (dismissed):
Hello @JoryHogeveen I think that in your PDF override, you setup a `$filename` property, so it works in your situation. But remember your PR aims to be merged in the Core used by 300.000 shops in the world that do not rely on your override 😅 so it has to work even without override.

I suggest a way to preserve backward compatibility
1. inject (so a new function) filename into PDF, so we can keep current naming convention
2. use $pdf->filename

Also in order to make it easier for QA analysts to verify your PR can you please submit a bug report? This information will make it easier for them to validate your contribution.

kpodemski commented:
Other places in the system have a hardcoded PDF filename, such as `OrderHistory`. I understand your change because `$pdf->render` gets the name based on the proper template, but we should do it globally if we want to change this.

JoryHogeveen commented:
> Other places in the system have a hardcoded PDF filename, such as `OrderHistory`. I understand your change because `$pdf->render` gets the name based on the proper template, but we should do it globally if we want to change this.

Sounds good to me, I'll update this PR to do so a.s.a.p.

JoryHogeveen commented:
@kpodemski @matks 

I've updated the PR to make sure invoice number formatting is only done in 1 location: OrderInvoice.
Also modified other references to reuse the PDF filename correctly.

matks commented on classes/pdf/PDF.php:
```diff
@@ -137,13 +137,6 @@ public function render($display = true)
-                }
-            }
-
             $template->assignHookData($object);
```
In previous implementation, when `assignHookData()` is called, `$pdf->filename` is set
In your implementation, when `assignHookData()` is called, `$pdf->filename` is not set

That could introduce issues for some modules.

Can I suggest we split the `$this->getFilename()` in two ? One `buildFilename()` and one `getFilename()`
```
    private function buildFilename(): string
    {
        foreach ($this->objects as $object) {
            $template = $this->getTemplateObject($object);
            if (!$template) {
                continue;
            }

            $filename = $template->getFilename();
            if (count($this->objects) > 1) {
                $filename = $template->getBulkFilename();
            }
        }

        return $filename;
    }
```

Then we do
`if (empty($this->filename)) {$this->filename = $this->buildFilename()}` at line 140.

But we can still use `$this->getFilename()` in other lines
```
    public function getFilename()
    {
        if (empty($this->filename)) {
            $this->filename = $this->buildFilename();
        }
    

        return $this->filename;
    }
```

JoryHogeveen replied:
I don't get why we would split it up. Why not just call `getFilename` just before assignHookData() ?
The method already returns the filename if it was already set.

JoryHogeveen replied:
Now that I think of it, why not generate the filename at __construct? I'd say this should always be available right?

matks replied:
> Now that I think of it, why not generate the filename at __construct? I'd say this should always be available right?
Indeed the sooner the better

JoryHogeveen replied:
Fixed in https://github.com/PrestaShop/PrestaShop/pull/26587/commits/bd03107cfd421d36383f19001062fe47ba15cd93

kpodemski commented:
@JoryHogeveen, you have some merge commits in your PR. Could you rebase it to get rid of these? Then it should be ok for me 👍

JoryHogeveen commented:
@kpodemski 
I removed my last commit. Sorry, I'm used to always pull changes from upstream (in this case branch 1.7.8.x).

Progi1984 commented on classes/pdf/PDF.php:
```diff
@@ -192,4 +187,30 @@ public function getTemplateObject($object)
+     *
+     * @return string
+     */
+    public function getFilename()
```
```suggestion
    public function getFilename(): string
```

Progi1984 commented on classes/pdf/PDF.php:
```diff
@@ -115,6 +115,8 @@ public function __construct($objects, $template, $smarty, $orientation = 'P')
             $this->send_bulk_flag = true;
         }
+
+        $this->getFilename();
```
Why do you call this method as you don't fetch the return value ?

JoryHogeveen replied:
The method sets the class property. This change was requested by @matks

matks replied:
@Progi1984 The method `getFilename()` sets the value of `$filename` internally.

Then `assignHookData()` is called which can use this variable.

If we **dont** call `getFilename()` at line 119 then
- before the PR: `$filename` was available when `assignHookData()` is called
- after the PR: `$filename` is not available when `assignHookData()` is called

Progi1984 replied:
In this case, the `getFilename` as a bad name as it define the variable `$filename`.

JoryHogeveen replied:
@Progi1984 ok so what do you want? Shall I also create a setter method?

Progi1984 commented on classes/pdf/HTMLTemplateInvoice.php:
```diff
@@ -501,17 +501,9 @@ public function getFilename()
-            $format,
-            Configuration::get('PS_INVOICE_PREFIX', $id_lang, null, $id_shop),
-            $this->order_invoice->number,
-            date('Y', strtotime($this->order_invoice->date_add))
-        ) . '.pdf';
+        return $invoiceNumber . '.pdf';
```
```suggestion
        return sprintf(
          '%s.pdf',
          $this->order_invoice->getInvoiceNumberFormatted($id_lang, $id_shop)
        );
```

JoryHogeveen replied:
Why would you do this? I see no benefit here.

Progi1984 replied:
We user sprintf to see fastly the final render of a string.

Progi1984 reviewed (commented):
@JoryHogeveen Thank you for your contribution. I have made some comments.

Progi1984 commented:
@JoryHogeveen As it's not a regression,  Could you target the `develop` branch, too ?

JoryHogeveen commented:
Hi @Progi1984 
I've tagged the 1.7.8.x branch because this PR is meant for 1.7.8.x and not a next major release. Otherwise I keep the need to manually update these changes on each update.

I've commented on your review comments. :)

JoryHogeveen commented:
@Progi1984 Any update on this? If possible it would be great if this could be merged into the next point releast!

JoryHogeveen commented:
Hi @Progi1984 

I did my best to accommodate anyone who had remarks on these changes and I believe it definitely enhances the usability of PS for developers but it's getting clear to me that it won't be for a while before such changes are even considered to be released.

If you do not want to merge this in `1.7.8.x` then please adjust this PR to suit your requirements for the `develop` branch yourself as I'm not sure what you would like to change and why.

MatShir commented:
Hi @JoryHogeveen, 
Thanks for your work and commitment, but it seems the PR fixes a feature (=improvement we considered as the same for analytics purposes, it's hard to distinguish feature to improvement sometimes). 
Meaning that is needs to be fixed into the develop branch. 

The 1.7.8.x is supported to fix only security issues, critical bugs and regressions (new bugs introduced by the 1.7.8.0). I hope you will understand, this process helps us to stabilize the software and assure of avoiding new regressions that can compromises merchant experience. 

We really want to see your contribution into PrestaShop, could you change your PR of branch ? We would be happy to review it :)

JoryHogeveen commented:
Hi @MatShir 

I understood that from @Progi1984 
I'll change the target branch now.
This PR is open and editable for anyone who can merge it so feel free to do with it what you want.

MatShir commented:
thanks 👌

sowbiba commented on classes/pdf/PDF.php:
```diff
@@ -192,4 +187,40 @@ public function getTemplateObject($object)
+     */
+    public function setFilename(): bool
+    {
+        foreach ($this->objects as $object) {
```
Thanks for the contribution @JoryHogeveen ;)

Maybe I dont understand this method purpose but, everytime an object has template, it overrides the filename.

However, You can put the count outside of the foreach and do
$this->filename = ($count>1) ? $template->getBulkFilename() : $template->getFilename();
or sthg like that

Finally I don't see why you return a bolean for this method. The return value is used nowhere

JoryHogeveen replied:
> Maybe I dont understand this method purpose but, everytime an object has template, it overrides the filename.

This only happens when no filename has been set yet.

> Finally I don't see why you return a bolean for this method. The return value is used nowhere

I added this since I saw it done in other setter methods as well, no other reason.

JoryHogeveen replied:
Fixed in https://github.com/PrestaShop/PrestaShop/pull/26587/commits/a6b421b9630f79353d29d67ae6d7ef96eb5a268b and https://github.com/PrestaShop/PrestaShop/pull/26587/commits/3a2aed27ff9a9c5906b84271a6a1c8152e1ce57b

atomiix commented on classes/pdf/PDF.php:
```diff
@@ -192,4 +187,40 @@ public function getTemplateObject($object)
+            $this->filename = $template->getFilename();
+            if (count($this->objects) > 1) {
+                $this->filename = $template->getBulkFilename();
+            }
+        }
+
+        return !empty($this->filename);
```
@JoryHogeveen I think @sowbiba is right, the loop misses a break or a return as soon as the name is defined.
```suggestion
            }
            if (!empty($this->filename)) {
                return true;
            }
        }

        return false;
```

JoryHogeveen replied:
Fixed in https://github.com/PrestaShop/PrestaShop/pull/26587/commits/a6b421b9630f79353d29d67ae6d7ef96eb5a268b

PierreRambaud reviewed (changes requested):
Unfortunately, this PR needs to be rebased :(

atomiix commented:
@JoryHogeveen rebase to the same branch to resolve the conflicted file `classes/PaymentModule.php`

PierreRambaud commented:
> @PierreRambaud Rebase to what?

We can't accept this kind of commits:
![image](https://user-images.githubusercontent.com/1462701/147356422-a2cc8691-4818-47b8-96b4-2bc087b616c8.png)


That's also the reason of the rebase instead of merge commits :)

JoryHogeveen commented:
Hi @PierreRambaud

I'm sorry but you'll have to specify what you want exactly, I contribute to many open source platforms and each of them do it differently.
The merge conflicts are fixed with a merge commit, that is the way I'm used to do such things.

Alternatively you can rebase yourself so you are sure it will be done correctly.

PierreRambaud commented:
> Hi @PierreRambaud
> 
> I'm sorry but you'll have to specify what you want exactly, I contribute to many open source platforms and each of them do it differently. The merge conflicts are fixed with a merge commit, that is the way I'm used to do such things.
> 
> Alternatively you can rebase yourself so you are sure it will be done correctly.

We use a GitHub action in order to deny "Merge commits", I know some open source project accept it, but in we won't do it.

You can remove your latest commit "Merge branch 'develop' into patch-1", and use the `git rebase` command instead.

JoryHogeveen commented:
> We use a GitHub action in order to deny "Merge commits", I know some open source project accept it, but in we won't do it.
> 
> You can remove your latest commit "Merge branch 'develop' into patch-1", and use the git rebase command instead.

Latest commit removed but `git rebase` does absolutely nothing. I get: `Current branch patch-1 is up to date.`

PierreRambaud commented:
> > We use a GitHub action in order to deny "Merge commits", I know some open source project accept it, but in we won't do it.
> > You can remove your latest commit "Merge branch 'develop' into patch-1", and use the git rebase command instead.
> 
> Latest commit removed but `git rebase` does absolutely nothing. I get: `Current branch patch-1 is up to date.`
> 
> EDIT: Ok, `git rebase develop` works but will take me hours because for some reason I need to manually accept each commit. If there is some easy (read: quick) way to do this I'm ok but at this rate, sorry, don't have the time.

There is one easy way, you can squash your commits and then use git rebase to rebase the only commit you have ^^
That's also why having a lot of commits that are not atomic isn't a good idea.

JoryHogeveen commented:
The reason it has so many commits is because each time the PR gets accepted by one dev, another dev is stepping in and requests changes..

Sorry but I'm out. Do with the PR what you want 👍

Progi1984 commented:
@JoryHogeveen Could you rebase your PR, please ?

PierreRambaud commented:
> The reason it has so many commits is because each time the PR gets accepted by one dev, another dev is stepping in and requests changes..
> 
> Sorry but I'm out. Do with the PR what you want +1

PR squashed and rebased.

khouloudbelguith reviewed (commented):
Hello @PrestaShop/prestashop-maintainers,

This issue https://github.com/PrestaShop/PrestaShop/issues/26646, which will be fixed by this PR, is not with the status ready.
It still needs specs.
Could you please check it?

Thank you!

Thanks!

MeKeyCool commented:
Hi @JoryHogeveen ! Thanks for your contribution and I feel sad that you were bored by the project process/guidelines.

As I am new to the project, I am currently discussing with @matks and @eternoendless about onboarding / contribution issues.
If you have any ideas / comments / critics please share. If you could share your feelings, some open source projects you enjoyed contributing and we can study, some "entry points" we should add to help contributors to understand the project guidelines, etc.

matks commented:
Thank you @JoryHogeveen 

@MeKeyCool tested the PR and validated it 👍
